src/main/scala/com/fulcrumgenomics/bam/SamRecordClipper.scala

```diff
@@ -0,0 +1,303 @@
+      Unit
+    }
+    else if (numberOfClippableBases(rec) <= numberOfBasesToClip) {
+      SAMUtils.makeReadUnmapped(rec)
+    }
+    else {
+      val oldElems = rec.getCigar.toSeq
+      val oldBases = rec.getReadBases
+      val oldQuals = rec.getBaseQualities
+      val (newElems, readBasesClipped, refBasesClipped, bases, quals) = clip(oldElems, numberOfBasesToClip, oldBases, oldQuals, mode)
+      rec.setAlignmentStart(rec.getAlignmentStart + refBasesClipped)
+      rec.setCigar(new Cigar(util.Arrays.asList(newElems:_*)))
+      rec.setReadBases(bases)
+      rec.setBaseQualities(quals)
+    }
+  }
+
+  /**
+    * Adds clipping of _at least_ numberOfBasesToClip to the end (right hand end)
+    * of an alignment. If clipping already exists at the end of the read it is
+    * preserved, and numberOfBasesToClip more clipping is added.
+    *
+    * If is unmapped or the numberOfBasesToClip is < 1 nothing is done.
+    *
+    * If the read has fewer clippable bases than requested clipping, the read is
+    * unmapped.
+    *
+    * If hard-clipping is requested and the read has existing soft-clipping at the end
+    * it is converted to hard-clipping before adding further clipping.
+    *
+    * If soft-clipping-with-masking is requested and the read already contains soft
+    * clipping at the end of the read, both the existing and new soft clipped bases
+    * are masked.
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to clip beyond any clipping
+    *                            already applied to the read
+    * @param mode the kind of clipping to be performed
+    */
+  def clipEndOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) : Unit = {
+    if (rec.getReadUnmappedFlag || numberOfBasesToClip < 1) {
+      Unit
+    }
+    else if (numberOfClippableBases(rec) <= numberOfBasesToClip) {
+      SAMUtils.makeReadUnmapped(rec)
+    }
+    else {
+      val oldElems = rec.getCigar.toSeq.reverse
+      val oldBases = rec.getReadBases.reverse
+      val oldQuals = rec.getBaseQualities.reverse
+      val (newElems, readBasesClipped, refBasesClipped, bases, quals) = clip(oldElems, numberOfBasesToClip, oldBases, oldQuals, mode)
+      rec.setCigar(new Cigar(util.Arrays.asList(newElems.reverse:_*)))
+      rec.setReadBases(bases.reverse)
+      rec.setBaseQualities(quals.reverse)
+    }
+  }
+
+  /**
+    * Attempts to clip an additional numberOfBasesToClip from the 5' end of the read. For
+    * details see [[clipStartOfAlignment()]] and [[clipEndOfAlignment()]].
+    *
+    * @param rec the record to be clipped
+    * @param numberOfBasesToClip the number of additional bases to be clipped
+    * @param mode the clipping mode
+    */
+  def clip5PrimeEndOfAlignment(rec: SAMRecord, numberOfBasesToClip: Int, mode: ClippingMode) = {
+    if (!rec.getReadUnmappedFlag && rec.getReadNegativeStrandFlag) clipEndOfAlignment(rec, numberOfBasesToClip, mode)
```

Comment:
Huh, ok.  I'm so used to SAMRecord pitching a fit if I try to retrieve a flag that it feels is in appropriate given the state of the read, I'm overly defensive.